prog/dftb+/lib_dftb/dftd4param.F90

```diff
@@ -0,0 +1,500 @@
+  pure function numIntegration(pol) result(trapzd)
+    implicit none
+    real(dp), intent(in) :: pol(23)
+    real(dp) :: trapzd
+
+    real(dp),parameter  :: freq(23) = [ &
+      & 0.000001_dp, 0.050000_dp, 0.100000_dp, 0.200000_dp, 0.300000_dp, &
+      & 0.400000_dp, 0.500000_dp, 0.600000_dp, 0.700000_dp, 0.800000_dp, &
+      & 0.900000_dp, 1.000000_dp, 1.200000_dp, 1.400000_dp, 1.600000_dp, &
+      & 1.800000_dp, 2.000000_dp, 2.500000_dp, 3.000000_dp, 4.000000_dp, &
+      & 5.000000_dp, 7.500000_dp, 10.00000_dp ]
+    !  just precalculate all weights and get the job done
+    real(dp),parameter :: weights(23) = 0.5_dp * [ &
+      & (freq (2) - freq (1)),  &
+      & (freq (2) - freq (1)) + (freq (3) - freq (2)), &
+      & (freq (3) - freq (2)) + (freq (4) - freq (3)), &
+      & (freq (4) - freq (3)) + (freq (5) - freq (4)), &
+      & (freq (5) - freq (4)) + (freq (6) - freq (5)), &
+      & (freq (6) - freq (5)) + (freq (7) - freq (6)), &
+      & (freq (7) - freq (6)) + (freq (8) - freq (7)), &
+      & (freq (8) - freq (7)) + (freq (9) - freq (8)), &
+      & (freq (9) - freq (8)) + (freq(10) - freq (9)), &
+      & (freq(10) - freq (9)) + (freq(11) - freq(10)), &
+      & (freq(11) - freq(10)) + (freq(12) - freq(11)), &
+      & (freq(12) - freq(11)) + (freq(13) - freq(12)), &
+      & (freq(13) - freq(12)) + (freq(14) - freq(13)), &
+      & (freq(14) - freq(13)) + (freq(15) - freq(14)), &
+      & (freq(15) - freq(14)) + (freq(16) - freq(15)), &
+      & (freq(16) - freq(15)) + (freq(17) - freq(16)), &
+      & (freq(17) - freq(16)) + (freq(18) - freq(17)), &
+      & (freq(18) - freq(17)) + (freq(19) - freq(18)), &
+      & (freq(19) - freq(18)) + (freq(20) - freq(19)), &
+      & (freq(20) - freq(19)) + (freq(21) - freq(20)), &
+      & (freq(21) - freq(20)) + (freq(22) - freq(21)), &
+      & (freq(22) - freq(21)) + (freq(23) - freq(22)), &
+      & (freq(23) - freq(22)) ]
+
+    trapzd = sum(pol*weights)
+
+  end function numIntegration
+
+
+  subroutine initializeCalculator(calculator, input, nAtom, izp)
+    use dftbp_constants, only: pi
+    type(DftD4Calculator), intent(inout) :: calculator
+    type(DispDftD4Inp), intent(in) :: input
+    !> Nr. of atoms (without periodic images)
+    integer, intent(in) :: nAtom
+    !> Atomic number of every atom.
+    integer, intent(in) :: izp(:)
+
+    integer :: mAt
+    integer :: iAt1, iZp1, iSec, iCN, iZp2, iRef1, iRef2
+    integer :: cncount(0:18)
+    real(dp) :: alpha(23), zEff1, c6, eta1
+    real(dp) :: tmp_hq(7, 118)
+
+    real(dp), parameter :: thopi = 3.0_dp/pi
+
+    ! initialize charge model
+    calculator%chi = chi
+    calculator%gam = gam
+    calculator%rad = rad
+    calculator%kcn = kcn
+
+    calculator%sqrtZr4r2 = sqrtZr4r2
+    calculator%covalentRadius = CovalentRadiusD3
+    calculator%electronegativity = paulingEN
+    calculator%ChemicalHardness = ChemicalHardness
+    calculator%EffectiveNuclearCharge = EffectiveNuclearCharge
+
```

Comment:
They currently map via the atomic number but could also use the species type instead, a copy allows to change this behaviour in one place. Also, I have encountered a lot of problems with OpenMP and global (constant) arrays in the past, since they seem to be handled inconsistently between compilers and OpenMP versions.

Reply:
OK.